src/execution/__tests__/oneof-test.ts

```diff
@@ -181,5 +186,338 @@ describe('Execute: Handles OneOf Input Objects', () => {
         ],
       });
     });
+
+    it('rejects a variable with multiple null keys', () => {
+      const query = `
+        query ($input: TestInputObject!) {
+          test(input: $input) {
+            a
+            b
+          }
+        }
+      `;
+      const result = executeQuery(query, rootValue, {
+        input: { a: null, b: null },
+      });
+
+      expectJSON(result).toDeepEqual({
+        errors: [
+          {
+            locations: [{ column: 16, line: 2 }],
+            message:
+              'Variable "$input" got invalid value { a: null, b: null }; Exactly one key must be specified for OneOf type "TestInputObject".',
+          },
+        ],
+      });
+    });
+
+    it('accepts a valid variable for field', () => {
+      const query = `
+        query ($a: String!) {
+          test(input: { a: $a }) {
+            a
+            b
+          }
+        }
+      `;
+      const result = executeQuery(query, rootValue, { a: 'abc' });
+
+      expectJSON(result).toDeepEqual({
+        data: {
+          test: {
+            a: 'abc',
+            b: null,
+          },
+        },
+      });
+    });
+
+    it('rejects multiple variables for fields', () => {
+      const query = `
+        query ($a: String!, $b: Int!) {
+          test(input: { a: $a, b: $b }) {
+            a
+            b
+          }
+        }
+      `;
+      const result = executeQuery(query, rootValue, { a: 'abc', b: 123 });
+
+      expectJSON(result).toDeepEqual({
+        data: {
+          test: null,
+        },
+        errors: [
+          {
+            // A nullable variable in a oneOf field position would be caught at validation-time
+            // hence the vague error message here.
+            message:
+              'Argument "input" of type "TestInputObject!" has invalid value { a: $a, b: $b }.',
```

Comment:
I understand the errors are vague and why; but the `{ a: $a }` error messages below are confusing. Maybe it would be better to replace all the errors with:

``Argument "input" of type "TestInputObject!" has invalid value - exactly one field must be specified, and the value for that field must not be `null`.``

Reply:
I updated the OneOf error messages to something uniform, but the vague ones are a result of the uniform behavior within `getArgumentValues()`

https://github.com/graphql/graphql-js/blob/78755526c8e89e2182c7f19bec68e36ca2279409/src/execution/values.ts#L232-L237 which does not transmit the original error (but would be caught at validation time).

Within #3813 , part of the rebased Default Value Coercion/Validation stack, @leebyron has fixed this to provide more informative errors.

Reply:
I think another way of saying what I am saying that all of the tests within the oneOf test file, including the new ones, actually belong within the tests for `coerceInputValue()` and `coerceInputLiteral`() (formerly known as `valueFromAST()`).

I'll set about to do that. Then these tests won't have opaque error messages in our test files that make it difficult to understand what exactly is being tested, although users who choose not to validate will still get opaque errors, at least until #3813 lands.

Reply:
This has been updated with the last couple of commits, and is now ready for re-review.

Reply:
Thinking about this more, moving the tests locations from file to file within this PR which ostensibly is just supposed to add a few new test cases will make the diff confusing. As, honestly, will updating the error messages as you requested. I should probably split this PR into multiple ones to better delineate what's happening.

@cito [has an open PR](https://github.com/graphql/graphql-js/pull/4181) to reduce the number of OneOf errors and I have submitted your suggestion within this PR to him to combine the error messages. So that PR can house the error message changes, and then this PR can be about adding the pre-coercion checks, and then we can do another one if necessary to move the tests around, which may in the end not be necessary.

Reply:
GitHub tip: if you reference PRs in a bulleted list then it will expand them showing their status and title; e.g. instead of:

> at least until #3813 lands.

You could write:

> at least until this PR lands:
>
> - #3813

I mention this because the "open PR" referenced above is now merged:

- #4181